system-addon/lib/TelemetryFeed.jsm

```diff
@@ -274,6 +280,37 @@ this.TelemetryFeed = class TelemetryFeed {
   }
 
   /**
+   * setPageModeForPerf - Set page mode for the session perf object. A page
+   * could be loaded in a combination of following modes: {
+   *   "regular": 0,
+   *   "preloaded": 1 << 0,
+   *   "prerendered": 1 << 1
+   * }
+   *
+   * Note:
+   *   1. Preloading doesn't apply to about:home.
+   *   2. The first newtab page is not preloaded, nor prerendered. This function
+   *      does not take that into account provided that it should not skew the
+   *      global measurement.
+   *
+   * @param {obj} session A session object
+   */
+  setPageModeForPerf(session) {
+    let pageMode = 0;
+
+    if (this._prefs.get("prerender")) {
+      pageMode |= PAGE_MODES_ENCODING.prerendered;
+    }
+
+    if (session.page !== "about:home" &&
+        Services.prefs.getBoolPref("browser.newtab.preload") &&
+        session.perf.topsites_first_painted_ts < session.perf.load_trigger_ts) {
+      pageMode |= PAGE_MODES_ENCODING.preloaded;
```

Comment:
This also feels susceptible to races.  One that jumps out at me is when preloading starts to happen, then the user triggers a load, then the first paint from the preload finishes.  I'm wondering if there's a way to fix this by using or adapting one of the [`tabbrowser.xml` attributes](http://searchfox.org/mozilla-central/source/browser/base/content/tabbrowser.xml#2194), but it's not behaving quite how I expect, so I'm not yet sure.

Reply:
Hmm, not sure, depends on how often that happens in practice. I believe that `session.perf.topsites_first_painted_ts < session.perf.load_trigger_ts` should cover the vast majority of preloading loads.

Reply:
Agreed, this isn't likely to be a huge problem, but if we can keep the data clean without too much work, that can save us pain and context-switching in the future.

Reply:
http://searchfox.org/mozilla-central/source/browser/base/content/tabbrowser.xml#2223 is actually the link I should have given above.

Reply:
https://bugzilla.mozilla.org/show_bug.cgi?id=1376895#c5 is where that isPreloaded attribute came from.  I suspect the easiest thing to do would be to talk to mconley about whether making that attribute persist (or change names?) rather than being removed.

Reply:
Though of course we'd still want to remove the attribute when the client navigates away from the page (there is already code in the tree to do that).